internal/service/cluster/discovery/join_peers.go

```diff
@@ -138,6 +142,57 @@ func dnsSRVResolver(opts Options, ctx context.Context) addressResolver {
 	})
 }
 
+const (
+	dnsSchemeDNS       = "dns+"
+	dnsSchemeDNSSRV    = "dnssrv+"
+	dnsSchemeDNSSRVNOA = "dnssrvnoa+"
+)
+
+// dnsSDURLResolver handles DNS-SD URLs which explicitly states what DNS query should be used for host resolve.
+//
+// Example: `dnssrv+_memcached._tcp.memcached.namespace.svc.cluster.local`
+//
+// Resolver rejects any non-URL values.
+func dnsSDURLResolver(opts Options, ctx context.Context) addressResolver {
+	srvLookup := opts.lookupSRVFn
+	if srvLookup == nil {
+		srvLookup = net.LookupSRV
+	}
+
+	return func(addr string) ([]string, error) {
+		var (
+			nextAddr     string
+			nextResolver addressResolver
+		)
+
+		switch {
+		case strings.HasPrefix(addr, dnsSchemeDNS):
+			nextAddr = addr[len(dnsSchemeDNS):]
+			nextResolver = dnsAResolver(opts, ctx)
+		case strings.HasPrefix(addr, dnsSchemeDNSSRV):
+			nextAddr = addr[len(dnsSchemeDNSSRV):]
+			nextResolver = dnsSRVResolver(opts, ctx)
+		case strings.HasPrefix(addr, dnsSchemeDNSSRVNOA):
+			nextAddr = addr[len(dnsSchemeDNSSRVNOA):]
+			nextResolver = dnsResolver(opts, ctx, "SRVNOA", func(addr string) ([]string, error) {
+				// NOTE: the only difference between SRVNOA and SRV, as SRV request should do N+1 query for A/AAAA.
+				_, addresses, err := srvLookup("", "", addr)
+				result := make([]string, 0, len(addresses))
+				for _, a := range addresses {
+					result = append(result, a.Target)
```

Comment:
Is there anything in spec about randomizing the order here? for clustering purposes, it may be useful to have the addresses in random order :)